src/librustc_lint/builtin.rs

```diff
@@ -1307,3 +1307,36 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+/// Lint for items marked `pub` that aren't exported to other crates
+pub struct UnexportedPub;
+
+declare_lint! {
+    UNEXPORTED_PUB,
+    Allow,
+    "`pub` items not reachable from crate root"
+}
+
+impl LintPass for UnexportedPub {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNEXPORTED_PUB)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnexportedPub {
+    fn check_item(&mut self, cx: &LateContext, item: &hir::Item) {
+        if !cx.access_levels.is_exported(item.id) && item.vis == hir::Visibility::Public {
+            let mut err = cx.struct_span_lint(UNEXPORTED_PUB, item.span, "unexported `pub` item");
+            err.help("consider exporting it for use by other crates");
+            // `pub` is three chars at start of declaration
```

Comment:
This is a bold claim!
`pub` can be passed to an item through a `$vis` matcher, for example.

Reply:
Could you add a test for this case? Something like
```rust
mod m {
    macro_rules! m {
        ($visibility: vis) => { $visibility struct S; }
    }

    m!(pub);
}
```